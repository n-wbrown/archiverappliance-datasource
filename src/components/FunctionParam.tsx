import React, { KeyboardEvent, FormEvent, FocusEvent } from 'react';
import Autosuggest from 'react-autosuggest';

export interface FunctionParamProps {
  param: string;
  paramDef: { name: string; options?: string[]; type: string };
  index: number;
  onChange: (paramIndex: number, newValue: string) => void;
  onRunQuery: () => void;
}

interface State {
  suggestions: any[];
  focused: boolean;
}

const getSuggestionValue = (suggestion: any) => {
  return suggestion;
};

const renderSuggestion = (suggestion: any) => {
  return <span>{suggestion}</span>;
};

class FunctionParam extends React.PureComponent<FunctionParamProps, State> {
  state = { suggestions: [], focused: false };

  constructor(props: FunctionParamProps) {
    super(props);
  }

  onChange = (paramIndex: number, event: FormEvent<any>, { newValue }: any) => {
    this.props.onChange(paramIndex, String(newValue));
  };

  onKeydownEnter = (event: KeyboardEvent<HTMLInputElement>) => {
    if (event.keyCode === 13) {
      event.currentTarget.blur();
    }
  };

  onPVSuggestionsFetchRequested = ({ value }: { value: any }, options: string[] | undefined) => {
    const suggestions = options === undefined ? [] : options;
    this.setState({
      suggestions: suggestions,
    });
  };

  onPVSuggestionsClearRequested = () => {
    this.setState({
      suggestions: [],
    });
  };

  onSuggestionsSelected = (
    event: React.FormEvent<any>,
    {
      suggestion,
      suggestionValue,
      suggestionIndex,
      sectionIndex,
      method,
    }: {
      suggestion: any;
      suggestionValue: string;
      suggestionIndex: number;
      sectionIndex: number | null;
      method: 'click' | 'enter';
    }
  ) => {
    if (method === 'enter') {
      event.currentTarget.blur();
    }
  };

  onFocus = (event: FocusEvent<any>) => {
    this.setState({
      focused: true,
    });
  };

  onBlur = (event: FocusEvent<any>) => {
    const { onRunQuery } = this.props;
    this.setState({
      focused: false,
    });
    onRunQuery();
  };

  calcInputWidth = (focused: boolean, param: string, placeholder: string) => {
    if (focused) {
      return '90px';
    }

    const paramLength = param.length;
    return paramLength < 1 ? `${placeholder.length}ch` : `${paramLength}ch`;
  };

  render() {
    const { suggestions, focused } = this.state;
    const { param, paramDef, index } = this.props;
    const inputWidth = this.calcInputWidth(focused, param, paramDef.name);
    return (
      <div className="aa-func-param">
        <Autosuggest
          suggestions={suggestions}
          onSuggestionsFetchRequested={e => this.onPVSuggestionsFetchRequested(e, paramDef.options)}
          onSuggestionsClearRequested={this.onPVSuggestionsClearRequested}
          onSuggestionSelected={this.onSuggestionsSelected}
          getSuggestionValue={getSuggestionValue}
          renderSuggestion={renderSuggestion}
          inputProps={{
            value: param,
            className: 'tight-form-func-param',
            placeholder: paramDef ? paramDef.name : '',
            spellCheck: false,
            onChange: (e, params) => this.onChange(index, e, params),
            onBlur: this.onBlur,
            onKeyDown: this.onKeydownEnter,
            onFocus: this.onFocus,
            style: { width: inputWidth, fontFamily: 'Consolas, "Courier New", Courier, Monaco, monospace' },
          }}
        />
      </div>
    );
  }
}

export { FunctionParam };
