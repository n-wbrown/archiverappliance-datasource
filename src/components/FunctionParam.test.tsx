import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { FunctionParam, FunctionParamProps } from './FunctionParam';

const setup = (propOverrides?: object) => {
  const props: FunctionParamProps = {
    param: 'avg',
    paramDef: {
      name: 'value',
      options: ['avg', 'min', 'max', 'absoluteMin', 'absoluteMax', 'sum'],
      type: 'string',
    },
    index: 1,
    onChange: jest.fn(),
    onRunQuery: jest.fn(),
  };

  Object.assign(props, propOverrides);

  return shallow(<FunctionParam {...props} />);
};

describe('Render', () => {
  it('should render component', () => {
    const wrapper = setup();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
