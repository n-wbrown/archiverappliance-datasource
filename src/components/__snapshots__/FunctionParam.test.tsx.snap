// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should render component 1`] = `
<div
  className="aa-func-param"
>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={true}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "className": "tight-form-func-param",
        "onBlur": [Function],
        "onChange": [Function],
        "onFocus": [Function],
        "onKeyDown": [Function],
        "placeholder": "value",
        "spellCheck": false,
        "style": Object {
          "fontFamily": "Consolas, \\"Courier New\\", Courier, Monaco, monospace",
          "width": "3ch",
        },
        "value": "avg",
      }
    }
    multiSection={false}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
    theme={
      Object {
        "container": "react-autosuggest__container",
        "containerOpen": "react-autosuggest__container--open",
        "input": "react-autosuggest__input",
        "inputFocused": "react-autosuggest__input--focused",
        "inputOpen": "react-autosuggest__input--open",
        "sectionContainer": "react-autosuggest__section-container",
        "sectionContainerFirst": "react-autosuggest__section-container--first",
        "sectionTitle": "react-autosuggest__section-title",
        "suggestion": "react-autosuggest__suggestion",
        "suggestionFirst": "react-autosuggest__suggestion--first",
        "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
        "suggestionsContainer": "react-autosuggest__suggestions-container",
        "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
        "suggestionsList": "react-autosuggest__suggestions-list",
      }
    }
  />
</div>
`;
